### PR TITLE
Complaint form module with all the changes

### PR DESCRIPTION
**Pull request by sshivam95**

### CODE REVIEW

File: app/src/main/java/app/insti/TagClass.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        this.color = getRandomColor();
+    }
+
+    public String getRandomColor() {
+        ArrayList<String> colors = new ArrayList<>();
+        colors.add("#ED7D31");
+        colors.add("#00B0F0");
+        colors.add("#FF0000");
+        colors.add("#D0CECE");
+        colors.add("#00B050");
```
Comment (pulsejet) [inline app/src/main/java/app/insti/TagClass.java]: All of these should be in resources

File: .idea/modules.xml [hunk 2]
```diff
@@ -2,8 +2,8 @@
 <project version="4">
   <component name="ProjectModuleManager">
     <modules>
-      <module fileurl="file://$PROJECT_DIR$/IITB-App.iml" filepath="$PROJECT_DIR$/IITB-App.iml" />
-      <module fileurl="file://$PROJECT_DIR$/IITBApp.iml" filepath="$PROJECT_DIR$/IITBApp.iml" />
+      <module fileurl="file://C:\Users\Shivam Sharma\AndroidStudioProjects\InstiApp\IITB-App.iml" filepath="C:\Users\Shivam Sharma\AndroidStudioProjects\InstiApp\IITB-App.iml" />
+      <module fileurl="file://C:\Users\Shivam Sharma\AndroidStudioProjects\InstiApp\IITBApp.iml" filepath="C:\Users\Shivam Sharma\AndroidStudioProjects\InstiApp\IITBApp.iml" />
```
Comment (pulsejet) [inline .idea/modules.xml]: Please don't change this

File: app/build.gradle [hunk 3]
```diff
@@ -38,7 +41,10 @@ dependencies {
     implementation "com.android.support:design:${supportLibVersion}"
     implementation "com.android.support:exifinterface:${supportLibVersion}"
     implementation "com.android.support:support-v4:${supportLibVersion}"
+    implementation "com.google.android.gms:play-services-gcm:${playServicesVersion}"
```
Comment (pulsejet) [inline app/build.gradle]: Why is this necessary?

File: app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public class CommentsViewHolder extends RecyclerView.ViewHolder {
+
+        private CardView cardView;
+        private CircleImageView circleImageView;
+        private TextView textViewName;
+        private TextView textViewCommentTime;
+        private TextView textViewComment;
+        final RetrofitInterface retrofitInterface = activity != null ? ((MainActivity) activity).getRetrofitInterface() : null;
```
Comment (pulsejet) [inline app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java]: MainActivity$getRetrofitInterface won't exist after #218

File: app/src/main/java/app/insti/adapter/ComplaintDetailsPagerAdapter.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public CharSequence getPageTitle(int position) {
+
+        if (position == 0) {
+            return "Complaint Details";
+        } else {
+            return "Complaint Details"; //Change the else as follows for Relevant COmplaints
+        }
```
Comment (pulsejet) [inline app/src/main/java/app/insti/adapter/ComplaintDetailsPagerAdapter.java]: Wth?

File: app/src/main/java/app/insti/adapter/ComplaintsRecyclerViewAdapter.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            }
+            try {
+                buttonComments.setText("COMMENTS(" + complaint.getComment().size() + ")");
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+            try {
+                buttonVotes.setText("UP VOTES(" + complaint.getUsersUpVoted().size() + ")");
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (pulsejet) [inline app/src/main/java/app/insti/adapter/ComplaintsRecyclerViewAdapter.java]: Put everything in one try - catch block instead of this and do null checks explicitly.

File: app/src/main/java/app/insti/api/ServiceGenerator.java [hunk 7]
```diff
@@ -116,4 +116,4 @@ public ServiceGenerator(Context mContext) {
     public RetrofitInterface getRetrofitInterface() {
         return retrofitInterface;
     }
-}
+}
```
Comment (pulsejet) [inline app/src/main/java/app/insti/api/ServiceGenerator.java]: Set your IDE to add newlines at the end of files automatically.

File: app/src/main/java/app/insti/api/response/ImageUploadResponse.java [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 import com.google.gson.annotations.SerializedName;
 
+import java.util.List;
```
Comment (pulsejet) [inline app/src/main/java/app/insti/api/response/ImageUploadResponse.java]: Why?

File: app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+        if (bundle != null) {
+            Log.i(TAG, "bundle != null");
+            callServerToGetDetailedComplaint();
+        }
+    }
+
+    private void upVote() {
+        if (voteCount == 0) {
+            RetrofitInterface retrofitInterface = ((MainActivity) getActivity()).getRetrofitInterface();
```
Comment (pulsejet) [inline app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java]: MainActivity$getRetrofitInterface will not exist after #218

File: app/src/main/java/app/insti/fragment/DetailedComplaintFragment.java [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+        }
+        try {
+            addVotesToView(detailedComplaint);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        try {
+            addCommentsToView(detailedComplaint);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (pulsejet) [inline app/src/main/java/app/insti/fragment/DetailedComplaintFragment.java]: Again, so many try catch blocks are ugly

File: app/src/main/java/app/insti/fragment/FeedFragment.java [hunk 11]
```diff
@@ -37,6 +38,7 @@
  */
 public class FeedFragment extends BaseFragment {
 
+    String TAG = "FeedFragment";
```
Comment (pulsejet) [inline app/src/main/java/app/insti/fragment/FeedFragment.java]: Why is this file changed at all?

File: app/src/main/res/layout/fragment_about.xml [hunk 12]
```diff
@@ -447,11 +439,6 @@
                 android:textStyle="bold" />
 
 
-
-
-
```
Comment (pulsejet) [inline app/src/main/res/layout/fragment_about.xml]: Again, why change this file at all? Possible conflict with #215

File: build.gradle [hunk 13]
```diff
@@ -20,8 +20,14 @@ buildscript {
 allprojects {
     repositories {
         jcenter()
-        maven { url 'https://maven.google.com' }
-        google()
```
Comment (pulsejet) [inline build.gradle]: Why?

File: build.gradle [hunk 14]
```diff
@@ -20,8 +20,14 @@ buildscript {
...
+        maven {
+            url 'https://maven.google.com/'
+            name 'Google'
+        }
+        maven {
+            url "https://jitpack.io"
+//            dont use proxy when syncing it
```
Comment (pulsejet) [inline build.gradle]: No place for a comment

File: app/src/main/java/app/insti/utils/TagCategories.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+            "Plastic Bottles (in Hostel messes)",
+            "Placards on Trees",
+            "Request for donation of clothes",
+            "Other donations",
+            "Cattle Issues",
+            "Autorickshaws",
+            "Potholes in Roads",
+            "Broken stormwater drains",
+            "Desilting - lakes",
+            "Flooding of roads and footpaths",
```
Comment (pulsejet) [inline app/src/main/java/app/insti/utils/TagCategories.java]: **IMPORTANT**
Why are these hard coded here? We cannot release an update if this list needs to be changed!

File: app/src/main/java/app/insti/activity/MainActivity.java [hunk 16]
```diff
@@ -562,6 +576,8 @@ public void updateFragment(Fragment fragment) {
             bundle.putString(Constants.USER_HOSTEL, session.isLoggedIn() && currentUser.getHostel() != null ? currentUser.getHostel() : "1");
         if (fragment instanceof SettingsFragment && session.isLoggedIn())
             bundle.putString(Constants.USER_ID, currentUser.getUserID());
+        if (fragment instanceof ComplaintFragment && session.isLoggedIn())
+            bundle.putString(Constants.USER_ID, currentUser.getUserID());
```
Comment (pulsejet) [inline app/src/main/java/app/insti/activity/MainActivity.java]: Could've just used a `||`

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: ok. But they won't be changing in the future so it doesn't matter

File: .idea/modules.xml (hunk 2 above)
Comment (sshivam95) [reply .idea/modules.xml]: The IDE does that

File: app/build.gradle (hunk 3 above)
Comment (sshivam95) [reply app/build.gradle]: This is for the Maps getting last known location

File: app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java (hunk 4 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java]: What is the solution for this?

File: app/src/main/java/app/insti/adapter/ComplaintDetailsPagerAdapter.java (hunk 5 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/adapter/ComplaintDetailsPagerAdapter.java]: I wanted to place a placeholder for "Relevant Complaints" for the next version

File: app/src/main/java/app/insti/adapter/ComplaintsRecyclerViewAdapter.java (hunk 6 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/adapter/ComplaintsRecyclerViewAdapter.java]: Will do

File: app/src/main/java/app/insti/api/ServiceGenerator.java (hunk 7 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/api/ServiceGenerator.java]: Ohk but will it matter?

File: app/src/main/java/app/insti/api/response/ImageUploadResponse.java (hunk 8 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/api/response/ImageUploadResponse.java]: Didn't realized that was there. Will do

File: app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java (hunk 9 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java]: Again, tell me the solution

File: app/src/main/java/app/insti/fragment/DetailedComplaintFragment.java (hunk 10 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/fragment/DetailedComplaintFragment.java]: ok. will put in single try catch

File: app/src/main/java/app/insti/fragment/FeedFragment.java (hunk 11 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/fragment/FeedFragment.java]: Added logs to check the flow of this fragment. Forgot to remove it. Sorry

File: app/src/main/res/layout/fragment_about.xml (hunk 12 above)
Comment (sshivam95) [reply app/src/main/res/layout/fragment_about.xml]: Didn't know that it will be changed after final merge

File: build.gradle (hunk 14 above)
Comment (sshivam95) [reply build.gradle]: Just a heads up for the person taking over after me

File: build.gradle (hunk 13 above)
Comment (sshivam95) [reply build.gradle]: Used for maps interface

File: app/src/main/java/app/insti/utils/TagCategories.java (hunk 15 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/utils/TagCategories.java]: We are working on the dynamic tags API for version 2. But for time being leave it as it is

File: app/src/main/java/app/insti/activity/MainActivity.java (hunk 16 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/activity/MainActivity.java]: ok. will do that

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/TagClass.java]: No, we can't rely on "being sure it won't change in the future"

File: .idea/modules.xml (hunk 2 above)
Comment (pulsejet) [reply .idea/modules.xml]: Don't commit it then!!

File: app/build.gradle (hunk 3 above)
Comment (pulsejet) [reply app/build.gradle]: Can you link documentation for that? I'm pretty sure gcm is referring to Google Cloud Messaging

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: It will conflict with the design styles of the app

File: app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java (hunk 4 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/adapter/CommentRecyclerViewAdapter.java]: Use `Utils$getRetrofitInterface` instead

File: app/src/main/java/app/insti/api/ServiceGenerator.java (hunk 7 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/api/ServiceGenerator.java]: It matters for two reasons:
1) Creates merge conflicts
2) The next time I add something below this (say) then the last line is changed unnecessary with a extra `\n`

File: app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java (hunk 9 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/fragment/ComplaintDetailsFragment.java]: `Utils$getRetrofitInterface`

File: build.gradle (hunk 13 above)
Comment (pulsejet) [reply build.gradle]: Not so sure these are not the same thing. @unstablebrainiac any thoughts on this? The last thing we want is a messed up gradle config.

File: build.gradle (hunk 14 above)
Comment (pulsejet) [reply build.gradle]: Not useful at all, since the next person would probably be working in a completely different environment, and this will only add to the confusion if something isn't working. On a side note, why are you using a proxy at all?

File: app/src/main/java/app/insti/utils/TagCategories.java (hunk 15 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/utils/TagCategories.java]: Put these in a JSON file and upload it as a static file to the server instead then. This really is no place for such a bunch of user-facing strings.

Comment (sshivam95) [reply app/src/main/java/app/insti/utils/TagCategories.java]: For time being, these are all temporary. Once the ML model kicks in, we won't be needing these. And I have to make an API for explicitly calling it on each character change.

File: build.gradle (hunk 13 above)
Comment (unstablebrainiac) [reply build.gradle]: This is unnecessay. Why do you even need JitPack? Have you tried compiling without these changes? If yes, did you face a problem? If yes, I'm sure we can come up with a better solution.

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/TagClass.java]: Make a new resources file with the colors and get those here. How would this conflict?

Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: Sajal said not to change anything in the colors.xml file

File: app/src/main/java/app/insti/utils/TagCategories.java (hunk 15 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/utils/TagCategories.java]: I won't pursue this then, but make sure these are proof read. Like I said, there _cannot_ be an update for a change in this.

File: build.gradle (hunk 13 above)
Comment (sshivam95) [reply build.gradle]: jitpack is required for syncing TagView and circleIndicator directly from github. Without jitpack, they won't sync.

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/TagClass.java]: Yeah, don't. Make a new xml file.

Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: ok

File: build.gradle (hunk 13 above)
Comment (pulsejet) [reply build.gradle]: [Agreed](https://stackoverflow.com/questions/46467561/difference-between-google-and-maven-url-https-maven-google-com).

File: app/src/main/java/app/insti/utils/TagCategories.java (hunk 15 above)
Comment (sshivam95) [reply app/src/main/java/app/insti/utils/TagCategories.java]: Yeah sure. Thanks. So I'll leave this as it is.

File: app/build.gradle (hunk 3 above)
Comment (pulsejet) [reply app/build.gradle]: Any heads up on this? Unnecessary dependencies have plagued this repo for too long.

Comment (sshivam95) [reply app/build.gradle]: Yeah, the gcm part is for cloud messaging, for maps it is play-services-map. I might have miswritten it

Comment (pulsejet) [reply app/build.gradle]: Great! Also check the other ones to be sure.

File: build.gradle (hunk 13 above)
Comment (sshivam95) [reply build.gradle]: So only google() is fine or nothing at all?

File: app/src/main/java/app/insti/api/RetrofitInterface.java [hunk 17]
```diff
@@ -107,4 +113,28 @@
...
+    @GET("complaints")
+    Call<List<Venter.Complaint>> getAllComplaints(@Header("Cookie") String sessionId);
+
+    @GET("complaints?filter=me")
+    Call<List<Venter.Complaint>> getUserComplaints(@Header("Cookie") String sessionId);
+
+    @GET("complaints/{complaintId}")
+    Call<Venter.Complaint> getComplaint(@Header("Cookie") String sessionId, @Path("complaintId") String complaintId);
+
+    @PUT("complaints/{complaintId}")
```
Comment (pulsejet) [inline app/src/main/java/app/insti/api/RetrofitInterface.java]: Note: Base URL has changed to `/api/venter` as of https://github.com/wncc/IITBapp/commit/ff12ca797401856dc43c51322d744597a6da220a

Comment (pulsejet) [reply app/src/main/java/app/insti/api/RetrofitInterface.java]: Refer to the [docs](https://wncc.github.io/IITBapp/) if there is any confusion.

File: app/src/main/java/app/insti/TagClass.java (hunk 1 above)
Comment (unstablebrainiac) [reply app/src/main/java/app/insti/TagClass.java]: I don't recall saying that

Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: Then is it fine if i add these in color.xml?

Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: Then is it fine if i add these in color.xml?

Comment (unstablebrainiac) [reply app/src/main/java/app/insti/TagClass.java]: I asked you to not use colours that don't match the app theme, and I'm still saying that. I don't know where you're using these colours as I've not gone through your code yet, but if you just moved these colours to a new file and are still using them as theme colours in your module, then you'll have to change this again.
I repeat, changing colors.xml is not a problem, not following the app theme is.

Comment (sshivam95) [reply app/src/main/java/app/insti/TagClass.java]: Ohk. Got it 👍

File: build.gradle (hunk 13 above)
Comment (pulsejet) [reply build.gradle]: Just don't change anything that doesn't **need** to be changed. Any refactor can be done separately.
